snuba/web/query.py

```diff
@@ -209,33 +268,69 @@ def log_query_and_update_stats(
     """
 
     stats.update(query_settings)
+
+    query_metadata.query_list.append(
+        ClickhouseQueryMetadata(
+            sql=sql, timer=timer, stats=stats, status=status, trace_id=trace_id,
+        )
+    )
+
+    return stats
+
+
+def record_query(
+    request: Request, timer: Timer, query_metadata: SnubaQueryMetadata
+) -> None:
     if settings.RECORD_QUERIES:
         # send to redis
-        state.record_query(
-            {
-                "request": request.body,
-                "sql": sql,
-                "timing": timer,
-                "stats": stats,
-                "status": status,
-            }
-        )
+        query_metadata.http_referrer = http_request.referrer
+        state.record_query(query_metadata.to_dict())
 
+        final = str(request.query.get_final())
+        referrer = request.referrer or "none"
         timer.send_metrics_to(
             metrics,
             tags={
-                "status": str(status),
-                "referrer": stats.get("referrer", "none"),
-                "final": str(stats.get("final", False)),
+                "status": query_metadata.status,
+                "referrer": referrer,
+                "final": final,
             },
-            mark_tags={"final": str(stats.get("final", False))},
+            mark_tags={"final": final},
         )
-    return stats
 
 
-@split_query
 def parse_and_run_query(
     dataset: Dataset, request: Request, timer: Timer
+) -> ClickhouseQueryResult:
+    """
+    Runs a query, then records the metadata about each split query that was run.
+    """
+    request_copy = copy.deepcopy(request)
+    query_metadata = SnubaQueryMetadata(
+        request=request_copy,
+        dataset=dataset,
+        timer=timer,
+        query_list=[],
+    )
```

Comment:
I don't think this can be realistically done earlier, but keep in mind this happens after the query has already been parsed into the AST. You may want to mention it in a comment.